### PR TITLE
system-probe: Use /proc/kallsyms to determine if eBPF will be supported by the current host

### PR DESCRIPTION
**Pull request by sfluor**
### What does this PR do?

Instead of checking the kernel version to determine if eBPF will be supported, directly read from `/proc/kallsyms` to see if the probes / functions we want to use are there.

We still have to read the kernel version to be able to exclude some know not-working versions.

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### CODE REVIEW

File: pkg/ebpf/common.go [hunk 1]
```diff
@@ -78,27 +79,52 @@ func verifyOSVersion(kernelCode uint32, platform string, exclusionList []string)
...
-		return false, fmt.Errorf(
-			"incompatible linux version. at least %s (%d) required, got %s (%d)",
-			kernelCodeToString(minRequiredKernelCode),
-			minRequiredKernelCode,
-			kernelCodeToString(kernelCode),
-			kernelCode,
-		)
+	kallsyms, err := readKallsyms()
+	if err != nil {
+		log.Warnf("error reading /proc/kallsyms file: %s", err)
```
Comment (shang-wang) [inline pkg/ebpf/common.go]: I guess this warning message is not totally accurate because the procRoot is not always `/proc/kallsyms` right?

Comment (shang-wang) [reply pkg/ebpf/common.go]: I guess warning on the return error should enough right?

File: pkg/ebpf/common.go [hunk 2]
```diff
@@ -78,27 +79,52 @@ func verifyOSVersion(kernelCode uint32, platform string, exclusionList []string)
...
+	for _, line := range lines {
+		fields := strings.Fields(line)
+		if len(fields) < 3 {
+			continue
+		}
+
+		funcs[fields[2]] = struct{}{}
+	}
+
+	for _, f := range requiredKernelFuncs {
```
Comment (shang-wang) [inline pkg/ebpf/common.go]: There might be a problem with this logic, if all lines have less than 3 fields, then your `funcs` is empty, then this `for` loop would not execute, and you returned `true` in the end, but in reality none of the lines contain `requiredKernelFuncs`, right?

File: pkg/ebpf/common.go (hunk 1 above)
Comment (sfluor) [reply pkg/ebpf/common.go]: 👍

File: pkg/ebpf/common.go (hunk 2 above)
Comment (sfluor) [reply pkg/ebpf/common.go]: 👍  I will change that

File: pkg/ebpf/common.go [hunk 3]
```diff
@@ -78,27 +79,52 @@ func verifyOSVersion(kernelCode uint32, platform string, exclusionList []string)
...
-
 	if isUbuntu(platform) {
 		if kernelCode >= linuxKernelVersionCode(4, 4, 119) && kernelCode <= linuxKernelVersionCode(4, 4, 126) {
 			return false, fmt.Errorf("got ubuntu kernel %s with known bug on platform: %s, see: https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1763454", kernelCodeToString(kernelCode), platform)
 		}
 	}
 
-	if kernelCode < minRequiredKernelCode {
-		return false, fmt.Errorf(
-			"incompatible linux version. at least %s (%d) required, got %s (%d)",
```
Comment (sunhay) [inline pkg/ebpf/common.go]: On failure to find file, can we also output this as a log error?

File: pkg/process/config/tracer_config.go [hunk 4]
```diff
@@ -55,18 +53,6 @@ func SysProbeConfigFromConfig(cfg *AgentConfig) *ebpf.Config {
 }
 
 func isIPv6EnabledOnHost() bool {
-	_, err := ioutil.ReadFile(filepath.Join(getProcRoot(), "net/if_inet6"))
+	_, err := ioutil.ReadFile(filepath.Join(util.GetProcRoot(), "net/if_inet6"))
```
Comment (sunhay) [inline pkg/process/config/tracer_config.go]: 👍

File: pkg/ebpf/common.go [hunk 5]
```diff
@@ -78,27 +79,52 @@ func verifyOSVersion(kernelCode uint32, platform string, exclusionList []string)
...
+	if err != nil {
+		log.Warnf("error reading /proc/kallsyms file: %s", err)
+		// If we can't read the /proc/kallsyms file let's just return true to avoid blocking the tracer from running
+		return true, nil
+	}
+
+	return verifyKernelFuncs(kallsyms), nil
+}
+
+func readKallsyms() (string, error) {
```
Comment (leeavital) [inline pkg/ebpf/common.go]: Spoke in person, and I think this should return a reader or scanner

File: pkg/ebpf/common.go (hunk 2 above)
Comment (sfluor) [reply pkg/ebpf/common.go]: actually this is already working (I added a test case for it though)

File: pkg/ebpf/common.go [hunk 6]
```diff
@@ -78,27 +79,52 @@ func verifyOSVersion(kernelCode uint32, platform string, exclusionList []string)
...
+		return "", errors.Wrapf(err, "error reading kallsyms file from proc dir: %s", procRoot)
 	}
 
-	return true, nil
+	return string(raw), nil
+}
+
+func verifyKernelFuncs(kallsyms string) bool {
+	lines := strings.Split(kallsyms, "\n")
+	funcs := make(map[string]struct{}, len(lines))
```
Comment (leeavital) [inline pkg/ebpf/common.go]: Can we write this without allocating such a large map?

I think something like:

```
found := map()
for _, requiredFunc := range required {
    found[requiredFunc] = false
}

for _, l := range lines {
   func := getFunc(l)
   if _, ok : =found[func] { found[func] = true }
}

// return true iff all entries in found are true
```

**Review by shang-wang — DISMISSED** [review]
Good for me

File: pkg/ebpf/common.go [hunk 7]
```diff
@@ -1,28 +1,30 @@
...
-	//                               4.1 - kprobes,
-	//                               4.4 - perf events)
-	// 	                      -> 4.4
-	minRequiredKernelCode = linuxKernelVersionCode(4, 4, 0)
-
-	// eBPF got backported to centOS
-	// see: https://www.redhat.com/en/blog/introduction-ebpf-red-hat-enterprise-linux-7
-	centOSMinRequiredKernelCode = linuxKernelVersionCode(3, 10, 0)
-)
+var requiredKernelFuncs = []string{
```
Comment (sunhay) [inline pkg/ebpf/common.go]: Can we keep the comment:`// Feature versions sourced from: https://github.com/iovisor/bcc/blob/master/docs/kernel-versions.md`

At the top of this?

**Review by sunhay — DISMISSED** [review]
One change, but looks great! Good idea on the `kallsyms`!

File: pkg/ebpf/common.go (hunk 7 above)
Comment (sfluor) [reply pkg/ebpf/common.go]: 👍

**Review by sunhay — APPROVED** [review]

**Pull request merged**